paasta_tools/cli/cmds/spark_run.py

```diff
@@ -0,0 +1,384 @@
+import json
+import os
+import socket
+import sys
+from random import randint
+
+from botocore.session import Session
+
+from paasta_tools.cli.cmds.check import makefile_responds_to
+from paasta_tools.cli.cmds.cook_image import paasta_cook_image
+from paasta_tools.cli.utils import check_if_port_free
+from paasta_tools.cli.utils import get_instance_config
+from paasta_tools.cli.utils import lazy_choices_completer
+from paasta_tools.cli.utils import list_instances
+from paasta_tools.cli.utils import list_services
+from paasta_tools.cli.utils import pick_random_port
+from paasta_tools.mesos_tools import find_mesos_leader
+from paasta_tools.mesos_tools import MESOS_MASTER_PORT
+from paasta_tools.utils import _run
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import get_username
+from paasta_tools.utils import load_system_paasta_config
+from paasta_tools.utils import NoConfigurationForServiceError
+from paasta_tools.utils import NoDeploymentsAvailable
+from paasta_tools.utils import NoDockerImageError
+from paasta_tools.utils import paasta_print
+from paasta_tools.utils import PaastaColors
+from paasta_tools.utils import PaastaNotConfiguredError
+from paasta_tools.utils import SystemPaastaConfig
+
+DEFAULT_SPARK_UI_PORT = 4040
+DEFAULT_SPARK_WORK_DIR = '/spark_client'
+DEFAULT_SPARK_DOCKER_IMAGE_PREFIX = 'paasta-spark-run'
+DEFAULT_SPARK_DOCKER_REGISTRY = 'docker-dev.yelpcorp.com'
+
+
+def add_subparser(subparsers):
+    list_parser = subparsers.add_parser(
+        'spark-run',
+        help="Run Spark on the PaaSTA cluster",
+        description=(
+            "'paasta spark-run' is useful for launching a Spark cluster on the PaaSTA mesos "
+            "cluster. It analyzes soa-configs and command line arguments to construct a "
+            "'docker run' invocation to match. By default the Spark service image is going "
+            "to pulled unless the --build option is used instead.\n\n"
+        ),
+    )
+    list_parser.add_argument(
+        '-b', '--build',
+        help="Build the docker image from scratch using the local Makefile's cook-image target.",
+        action='store_true',
+        default=False,
+    )
+
+    list_parser.add_argument(
+        '-s', '--service',
+        help="The name of the service from which the Spark image is built.",
+        default='spark',
+    ).completer = lazy_choices_completer(list_services)
+
+    list_parser.add_argument(
+        '-i', '--instance',
+        help=("Start a docker run for a particular instance of the service."),
+        default='client',
+    ).completer = lazy_choices_completer(list_instances)
+
+    # Restrict usage to norcal-devc and pnw-devc for now.
+    list_parser.add_argument(
+        '-c', '--cluster',
+        help=(
+            "The name of the cluster you wish to run Spark on. "
+            "If omitted, uses the default cluster defined in the paasta spark-run configs."
```

Comment:
It's not clear where the default cluster is defined. (Is this in yelpsoa-configs?)

Reply:
Eventually we may have spark_run_config['default_cluster']. At the moment, I wanted to limit usage on the two devc clusters and made --cluster required. I think this help remind users which mesos UI to look for their framework info.